### PR TITLE
feat: CDP Templates

### PR DESCRIPTION
**Pull request by benjackwhite**
## Problem

Given that people can write their own functions, we don't need to offer templates from the DB, but we do want to still have some control over them to make development way faster (automated tests etc.)

## Changes

* Adds a templates area (for now in python) for building and testing all integrations
* Adds a duplicate button for the relevant hog function
* Indicates when built from a template with options for resetting, duplicating etc.
* Fixes a bunch of issues around creation failures
* Adds automatic icon suggestions

## TODO
- [x] Save the template ID against the hog function if created from one
- [x] Indicate if the function has deviated from the template with an option to reset it to the template
- [x] Add a bunch more templates (slack etc)

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

### CODE REVIEW

File: posthog/cdp/templates/README.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+1. Easier to manage as they are just code in the `posthog` repo - we don't need to worry about keeping a database in sync
+2. Update conflicts are left to the user to manage - changing the template doesn't change their function. It only will indicate in the UI that it is out of sync and they can choose to pull in the updated template changes, resolving any issues at that point.
+3. Sharing templates becomes very simple - it can all be done via a URL
+
+## Notes on building good templates
+
+Templates should be as generic as possible. The underlying Hog code should only do what is required for communicating with the destination or triggering the workflow.
+
+All input data should be controlled via the `inputs_schema` wherever possible as this leaves ultimate flexibility in the hands of the user of the template as they can inject data from the source (whether it is an Event, an ActivityLog or anything else) using Hog templating. If you hit the limits of the input templating, considering working with #team-cdp to extend these before writing a Hog function that is too tightly coupled to
```
Comment (mariusandra) [inline posthog/cdp/templates/README.md:13]: it feels wise to finish sentences

File: posthog/cdp/templates/slack/template_slack.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    id="template-slack",
+    name="Slack webhook",
+    description="Sends a webhook templated by the incoming event data",
+    icon_url="/api/projects/@current/hog_functions/icon/?id=slack.com",
+    hog="""
+fetch(inputs.url, {
+  'headers': inputs.headers,
+  'body': inputs.body,
+  'method': inputs.method
+});
```
Comment (mariusandra) [inline posthog/cdp/templates/slack/template_slack.py:15]: Time to start removing all these semicolons

**Review by mariusandra — APPROVED** [review]

File: frontend/src/scenes/pipeline/hogfunctions/pipelineHogFunctionConfigurationLogic.tsx [hunk 3]
```diff
@@ -204,13 +237,29 @@ export const pipelineHogFunctionConfigurationLogic = kea<pipelineHogFunctionConf
...
     listeners(({ actions, values, cache, props }) => ({
-        loadTemplateSuccess: () => actions.resetForm(),
-        loadHogFunctionSuccess: () => actions.resetForm(),
-        resetForm: () => {
-            const savedValue = values.hogFunction ?? values.template
+        loadTemplateSuccess: ({ template }) => {
+            // Fill defaults from template
+            const inputs = {}
+
+            template!.inputs_schema?.forEach((schema) => {
```
[the remark this answers is not in the document]
Comment (benjackwhite) [reply frontend/src/scenes/pipeline/hogfunctions/pipelineHogFunctionConfigurationLogic.tsx:244]: Fair. I should just have a quick exit check

File: posthog/cdp/templates/README.md (hunk 1 above)
Comment (benjackwhite) [reply posthog/cdp/templates/README.md:13]: dad joke ;)

File: frontend/src/scenes/pipeline/hogfunctions/pipelineHogFunctionConfigurationLogic.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (benjackwhite) [reply frontend/src/scenes/pipeline/hogfunctions/pipelineHogFunctionConfigurationLogic.tsx:244]: I'll fix in a follow up

**Pull request merged**